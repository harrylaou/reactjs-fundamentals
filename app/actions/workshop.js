export const RECEIVE_WORKSHOP = 'RECEIVE_WORKSHOP'


export const receiveWorkshop = workshop => ({
  type: RECEIVE_WORKSHOP,
  workshop
})