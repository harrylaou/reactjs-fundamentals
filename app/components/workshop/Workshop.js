import React, {Component} from 'react'
import {
  Card,
  CardActions,
  CardHeader,
  CardMedia,
  CardTitle,
  CardText
} from 'material-ui/Card'
import es6promise from 'es6-promise'
import 'isomorphic-fetch'

es6promise.polyfill()

const Workshop = props => {

  let workshop = props.workshop
  let id = props.id
  console.log("Workshop id", props.id)

  if (workshop && workshop.id != id) {
    props.fetchWorkshop(id)
  }
  let title = ''
  let price = ''
  let url = ''
  let avatar = ''

  if (workshop) {
    title = `${workshop.title}`
    price = `${workshop.price}`
    url = `${workshop.url}`
    avatar = `${workshop.instructors[0].avatar}`
  }

  return (
    <Card>
      <CardHeader title={title} subtitle={url} avatar={avatar}/>
      <CardActions/>
    </Card>
  )
}

export default Workshop
