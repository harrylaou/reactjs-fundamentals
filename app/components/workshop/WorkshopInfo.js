import React, {Component} from 'react'
import {
  Card, CardActions, CardHeader, CardTitle
} from 'material-ui/Card'
import FlatButton from 'material-ui/FlatButton'

const WorkshopInfo = (props) => {
  let workshop = props.workshop
  let workshopID = props.workshopID
  let instructors = ''
  let title = ''
  let button = ''

  if (workshop && workshop.id != workshopID) {
    props.fetchWorkshop(workshopID)
  }

  if (workshop) {
    instructors = workshop.instructors.map((instructor, index) => (
      <CardHeader
        key={index}
        title={instructor.name}
        subtitle={instructor.url}
        avatar={instructor.avatar}
      />
    ))
    title = `${workshop.title}  -  ${workshop.price}`
    button = (<FlatButton label="Buy" href={workshop.url} />)
  }

  return (
    <Card>
      <CardTitle title={title} />
      <CardTitle title="Instructors" />
      {instructors}
      <CardActions>
        {button}
      </CardActions>
    </Card>
  )
}

WorkshopInfo.propTypes = {
  workshop: React.PropTypes.object,
  workshopID: React.PropTypes.string
}

export default WorkshopInfo
