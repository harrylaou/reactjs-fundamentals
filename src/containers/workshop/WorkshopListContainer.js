import React, {Component} from 'react'
import {connect} from 'react-redux'
import WorkshopList from '../../components/workshop/WorkshopList'
import * as api from '../../api'
import * as actions from '../../actions/workshops'
import withWidth from '../../utils/WithWidth'
import { withRouter } from 'react-router'

class WorkshopListContainer extends Component {
  constructor() {
    super()
    this.showWorkshop = this.showWorkshop.bind(this)
  }

  componentDidMount() {
    // Imagine we want to fetch workshops from another component. How do you
    // think we could reuse this code?
    api
      .getWorkshops()
      .then((workshops) => {
        this.props.dispatch(actions.receiveWorkshops(workshops))
      })
      .catch((err)=> {
        console.log(err)
      })
  }

  showWorkshop(workshop) {
    this.props.router.push(`/workshops/${workshop.id}`)
  }

  render() {
    return (
      <WorkshopList
        {...this.props}
        workshop={this.props.params.workshop}
        workshops={this.props.workshops}
        showWorkshop={this.showWorkshop}
        width={this.props.width}
      />
    )
  }
}

WorkshopListContainer.propTypes = {
  width: React.PropTypes.number.isRequired,
  router: React.PropTypes.object.isRequired,
  workshops: React.PropTypes.array,
  dispatch: React.PropTypes.func.isRequired,
  params: React.PropTypes.object.isRequired
}

const mapStateToProps = (state) => ({
  workshops: state.workshops
})

// Could you refactor this to make it better? Hint: LoginContainer
const mapDispatchToProps = (dispatch) => ({
  dispatch
})

// Do you think the order of these components matter?
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(withWidth(WorkshopListContainer)))
