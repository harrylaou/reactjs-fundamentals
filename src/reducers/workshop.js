import { RECEIVE_WORKSHOP } from '../actions/workshop'

const workshop = (state = null, action) => {
  switch (action.type) {
    case RECEIVE_WORKSHOP: {
      return action.workshop
    }
    default:
      return state
  }
}

export default workshop
