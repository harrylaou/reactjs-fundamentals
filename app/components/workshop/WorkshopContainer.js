import React, {Component} from 'react'

import es6promise from 'es6-promise'
import 'isomorphic-fetch'
import Workshop from './Workshop'

es6promise.polyfill()

class WorkshopContainer extends Component {
  constructor() {
    super()
    this.fetchWorkshop = this.fetchWorkshop.bind(this)
    this.state = { id: null }
  }

  componentDidMount() {
    console.log("WorkshopContainer id",this.props.params.id)
    this.fetchWorkshop(this.props.params.id)
  }

  fetchWorkshop(id) {
    fetch(`/data/workshops/${id}.json`, {
        method: 'get'
    }).then((response) => {
        return response.json()
    }).then((data) => {
        this.setState({workshop : data})
    }).catch((err)=> {
        console.log(err)
    })
  }

  render() {

    return (
    <Workshop
        id={this.props.params.id}
        workshop={this.state.workshop}
        fetchWorkshop={this.fetchWorkshop}
    />
    )
  }
}

export default WorkshopContainer
