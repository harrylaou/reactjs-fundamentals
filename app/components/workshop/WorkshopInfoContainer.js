import React, {Component} from 'react'
import WorkshopInfo from './WorkshopInfo'
import es6promise from 'es6-promise'
import 'isomorphic-fetch'

es6promise.polyfill()

class WorkshopInfoContainer extends Component {
  constructor() {
    super()
    this.state = { workshop: null }
    this.fetchWorkshop = this.fetchWorkshop.bind(this)
  }

  componentDidMount() {
    this.fetchWorkshop(this.props.params.workshop)
  }

  fetchWorkshop(workshop) {
    fetch(`/data/workshops/${workshop}.json`, {
      method: 'get'
    }).then((response) => {
      return response.json()
    }).then((data) => {
      this.setState({workshop : data})
    }).catch((err)=> {
      console.log(err)
    })
  }

  render() {
    return (
      <WorkshopInfo
        {...this.props}
        fetchWorkshop={this.fetchWorkshop}
        workshop={this.state.workshop}
        workshopID={this.props.params.workshop}
      />
    )
  }
}

export default WorkshopInfoContainer