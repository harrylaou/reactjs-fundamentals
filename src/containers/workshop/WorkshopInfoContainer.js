import React, {Component} from 'react'
import {connect} from 'react-redux'
import WorkshopInfo from '../../components/workshop/WorkshopInfo'
import * as api from '../../api'
import * as actions from '../../actions/workshop'

class WorkshopInfoContainer extends Component {
  constructor() {
    super()
    this.fetchWorkshop = this.fetchWorkshop.bind(this)
  }

  componentDidMount() {
    this.fetchWorkshop(this.props.params.workshop)
  }

  componentWillReceiveProps(nextProps) {
    // If you are thinking that performance could be improved by caching
    // the user, we agree. But we want to keep this example simple.
    // We are going to improve this code in the next lessons. We are building up :)
    if (this.props.params.workshop !== nextProps.params.workshop) {
      this.fetchWorkshop(nextProps.params.workshop)
    }
  }

  // Imagine we want to fetch another workshop from another component. How do you
  // think we could reuse this code?
  fetchWorkshop(workshop) {
    api
      .getWorkshop(workshop)
      .then((workshop) => {
        this.props.dispatch(actions.receiveWorkshop(workshop))
      })
      .catch((err)=> {
        console.log(err)
      })
  }

  render() {
    return (
      <WorkshopInfo
        {...this.props}
        workshop={this.props.workshop}
      />
    )
  }
}

WorkshopInfoContainer.propTypes = {
  params: React.PropTypes.object.isRequired,
  dispatch: React.PropTypes.func.isRequired,
  workshop: React.PropTypes.object
}

const mapStateToProps = (state) => ({
  workshop: state.workshop
})

// Could you refactor this to make it better? Hint: LoginContainer
const mapDispatchToProps = (dispatch) => ({
  dispatch
})

// Do you think the order of these components matter?
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(WorkshopInfoContainer)
