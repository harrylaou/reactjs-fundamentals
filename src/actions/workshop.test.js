import {receiveWorkshop, RECEIVE_WORKSHOP} from './workshop'

test(`Return an Action tyoe ${RECEIVE_WORKSHOP}`, () => {
  const workshop = "Reactjs Fundamentals"
  expect(receiveWorkshop(workshop)).toEqual({
    type: RECEIVE_WORKSHOP,
    workshop
  })

  expect(receiveWorkshop(workshop)).toMatchSnapshot()
})