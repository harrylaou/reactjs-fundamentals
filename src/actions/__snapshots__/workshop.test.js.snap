// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Return an Action tyoe RECEIVE_WORKSHOP 1`] = `
Object {
  "type": "RECEIVE_WORKSHOP",
  "workshop": "Reactjs Fundamentals",
}
`;
